cephadm: add RGW SSL support

matthewoliver opened the pull request:
This patch adds SSL support to RGW when using cephadm.
If an SSL certificate is provided, inside the json supplied with:
    
  cpeh orchestrator rgw create -i rgw.json
    
Then the SSL cert and/or key will be added to pushed into the mon config-key database
using the key `rgw/cert/<rgw_realm>/<rgw_zone>.[crt|key]`.
Which will then be referenced in the config:
    
  rgw_frontends = beast port=80 ssl_port=443 ssl_certificate=config://rgw/cert/<rgw_realm>/<rgw_zone>.crt
    
And if an ssl key is also supplied this becomes something like:
    
  rgw_frontends = beast port=80 ssl_port=443 ssl_certificate=config://rgw/cert/<rgw_realm>/<rgw_zone>.crt ssl_key=config://rgw/cert/<rgw_realm>/<rgw_zone>.crt
    
Of course you could also just upload the cert and key yourself to
config-key location, and ssl will be enabled as well. But this patch
let's you either supply them via `-i` or as a manual upload step.

Co-Authored-By: Michael Fritch <mfritch@suse.com>
Co-Authored-By: Sebastian Wagner <sebastian@spawnhost.de>
Signed-off-by: Matthew Oliver <moliver@suse.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

sebastian-philipp commented:
jenkins test make check

sebastian-philipp commented on src/pybind/mgr/cephadm/module.py:
```diff
@@ -1045,7 +1045,7 @@ def remove_osds(self, osd_ids, services):
     def _create_daemon(self, daemon_type, daemon_id, host, keyring,
                        extra_args=None, extra_config=None,
-                       reconfig=False):
+                       reconfig=False, files=None):
```
```suggestion
        def _create_daemon(self,
            daemon_type,  # type: str
            daemon_id,  # type: str 
            host,  # type: str 
            keyring,  # type: str
            extra_args=None,  # type: Optional[List[str]] 
            extra_config=None,  # type: ????
            reconfig=False,  # type: bool 
            files=None  # type: Optional[Dict[str, str]]
        ):
```
(that would be some form of documentation)

matthewoliver replied:
How do we represent the files being either a Dict[str, str] or Dict[str, List[str]] ? Or should we just always assume the latter? Because they could be a combination. I use an isinstance(..., (list, tuple)). and turn it into 1 string in the list case.

sebastian-philipp commented on src/pybind/mgr/orchestrator.py:
```diff
@@ -1221,15 +1221,15 @@ class StatelessServiceSpec(object):
 
     def __init__(self, name, placement=None):
-        self.placement = PlacementSpec() if placement is None else placement  # type: PlacementSpec
+        self.placement = PlacementSpec(count=1) if placement is None else placement  # type: PlacementSpec
```
My experience is, the orchestrator.py is the wrong place to invent _any_ defaults. Instead if we need defaults, we should do them in the cephadm/module.py module.

matthewoliver replied:
OK, fair enough. However, line 1232's `else 1` never fires, and we have some defaulting going on. And then I tried to use -i in my vstart it would fail with a count being null. So just moved to it up here. Placement object actually gets recreated in the RGW spec after this anyway... so maybe more clean up is in order.

sebastian-philipp commented on src/pybind/mgr/orchestrator.py:
```diff
@@ -1272,7 +1272,10 @@ def __init__(self,
+                 count=None,  # type: Optional[int]
+                 rgw_frontend_ssl_port=443,  # type: Optional[int]
+                 rgw_frontend_ssl_certificate_path="/etc/ceph/rgw.pem",  # type: Optional[str]
+                 rgw_frontend_ssl_certificate=None   # type: Optional[str]
```
Again, :-1: for inventing _any_ defaults here.

matthewoliver replied:
Yup good point. I'll remove defaults and just make them None.

sebastian-philipp commented on src/pybind/mgr/orchestrator_cli/module.py:
```diff
@@ -466,13 +466,13 @@ def _mds_rm(self, name):
-        'name=realm_name,type=CephString '
-        'name=zone_name,type=CephString '
+        'name=realm_name,type=CephString,req=false '
+        'name=zone_name,type=CephString,req=false '
```
Last week, I stumbled across this, too. Especially as it is non-obvious that you should probably call this with `default` `default`. :+1: from my side here, except that we'Re now creating an imbalance between `add` and `rm`, as `rm` still requires them.

What about just marking `realm` as `req=false`?

matthewoliver replied:
but then your still sending in one 'default' and we're suppose to be fully defining the RGW in json. So it's redundant. I agree the imbalance between `add` and `update` or 'rm', did make me question this. But then figured if only add supports `inbuf` or `(-i)` then that's why `req=false` is required.

Of course we could change the useage to:

```
usage = """
Usage:
  ceph orchestrator rgw add <zone_name> -i <json_file>
  ceph orchestrator rgw add <realm_name> <zone_name>
    """
```
But that just doesn't look right.

sebastian-philipp commented on src/pybind/mgr/orchestrator_cli/module.py:
```diff
@@ -484,10 +484,14 @@ def _rgw_add(self, realm_name, zone_name, num=1, hosts=None, inbuf=None):
-            rgw_realm=realm_name,
-            rgw_zone=zone_name,
-            placement=orchestrator.PlacementSpec(hosts=hosts, count=num))
+        elif realm_name and zone_name:
```
`zone_name is not None and realm_name is None` should work as well, right?

matthewoliver replied:
Sure, `if variable:` covers both `None` and `''` both of which are illegal in this case. But sure, happy to change this if we only want to check for not None.

mgfritch commented on src/cephadm/cephadm:
```diff
@@ -848,7 +848,7 @@ def get_config_and_keyring():
 
-def get_config_and_both_keyrings():
+def get_config_keyrings_and_files():
     # type: () -> Tuple[str, str, Optional[str]]
```
```suggestion
     # type: () -> Tuple[str, str, Optional[str], Dict]
```

matthewoliver replied:
done

mgfritch commented on src/cephadm/cephadm:
```diff
@@ -860,7 +860,14 @@ def get_config_and_both_keyrings():
+        if d.get('keyring'):
+            keyring = d.pop('keyring')
+        if d.get('crash_keyring'):
+            crash_keyring = d.pop('crash_keyring')
```
```suggestion
        config = d.pop('config', '')
        keyring = d.pop('keyring', '')
        crash_keyring = d.pop('crash_keyring', '')
```

mgfritch commented on src/cephadm/cephadm:
```diff
@@ -971,10 +978,27 @@ def extract_uid_gid(img='', file_path='/var/lib/ceph'):
     return (int(uid), int(gid))
 
+def deploy_extra_files(extra, data_dir, uid, gid):
+    # type: (dict[str, list[str]], str, int, int) -> dict[str, str]
```
```suggestion
    # type: (Dict[str, List[str]], str, int, int) -> Dict[str, str]
```

cbodley commented:
very cool! does this also make it possible to run ssl-only? ie, if you omit `rgw_frontend_port` but provide a `rgw_frontend_ssl_port`, do we still bind to port=80?

mgfritch commented on src/cephadm/cephadm:
```diff
@@ -971,10 +978,27 @@ def extract_uid_gid(img='', file_path='/var/lib/ceph'):
+            with open(host_filename, 'w') as f:
+                os.fchown(f.fileno(), uid, gid)
+                os.fchmod(f.fileno(), 0o600)
+                f.write(data)
```
```
error: Incompatible types in assignment (expression has type "str", variable has type "List[str]")
```

Build a str `data_str` from the List `data` ... ?

```suggestion
            data_str = ''
            if isinstance(data, (list, tuple)):
                data_str = os.pathsep.join(data)
            makedirs(os.path.dirname(host_filename), uid, gid, 0o755)
            with open(host_filename, 'w') as f:
                os.fchown(f.fileno(), uid, gid)
                os.fchmod(f.fileno(), 0o600)
                f.write(data_str)
```

mgfritch commented on src/cephadm/cephadm:
```diff
@@ -971,10 +978,27 @@ def extract_uid_gid(img='', file_path='/var/lib/ceph'):
+            host_filename = os.path.basename(filename)
+            host_filename = os.path.join(data_dir, 'extra', host_filename)
+            if isinstance(data, (list, tuple)):
+                data = "\n".join(data)
```
```
error: Incompatible types in assignment (expression has type "str", variable has type "List[str]")
```

Maybe deal with the line endings while doing the file write?

```suggestion
        # There are extra files to place in the data dir
        for filename, data in extra.items():
            host_filename = os.path.basename(filename)
            host_filename = os.path.join(data_dir, 'extra', host_filename)
            makedirs(os.path.dirname(host_filename), uid, gid, 0o755)
            with open(host_filename, 'w') as f:
                os.fchown(f.fileno(), uid, gid)
                os.fchmod(f.fileno(), 0o600)
                for line in data:
                    if not line.endswith(os.linesep):
                        line += os.linesep
                    f.write(line)
            mounts[host_filename] = filename
```

matthewoliver replied:
From what I've read, Python is smart enough to use the correct line endings so we only need '\n'. And I assumed if the file has line endings we want them retained (think things like authorised_hosts etc). If there wasn't suppose to be any line endings, then we'd either have 1 string in the list or just a string (which is what the isinstance(.., (list, tuple)) is catching.

matthewoliver replied:
Aslo, ok so we're doing tests in string typing, so can't just reuse a variable and change it's type, good to know, I can fix these :)

liewegas commented:
Two high-level comments:
1. I don't think we should be putting anything in the daemon's ceph.conf unless it is absolutely necessary.  Instead, we should put options in the mon's config database, and (in most cases) apply them to the entire set/zone of rgw daemons.  This puts config mostly out of scope for the orchestrator layer.
2. IMO the SSL cert should be stored on the mon's config-key database.  We should teach radosgw how to fetch it from there when it starts up instead of copying it onto each node.

matthewoliver commented:
> very cool! does this also make it possible to run ssl-only? ie, if you omit `rgw_frontend_port` but provide a `rgw_frontend_ssl_port`, do we still bind to port=80?

I'd have to test, but when we build up the `rgw_frontends` config if port is None then it wont add it. So assume it wont, ie create:
`  rgw_frontends = beast ssl_port=443 ssl_certificate=/etc/ceph/rgw.pem`

matthewoliver commented:
> Two high-level comments:
> 
> 1. I don't think we should be putting anything in the daemon's ceph.conf unless it is absolutely necessary.  Instead, we should put options in the mon's config database, and (in most cases) apply them to the entire set/zone of rgw daemons.  This puts config mostly out of scope for the orchestrator layer.

I thought I was adding everything to the mon config database. I don't touch the ceph.conf added to the container. Though I'm probably just misunderstanding what your saying.

> 2. IMO the SSL cert should be stored on the mon's config-key database.  We should teach radosgw how to fetch it from there when it starts up instead of copying it onto each node.

oh interesting to put it in via `ceph config ...` and then in rgw look for the cert data there. That's kinda cool. How about if RGW sees the `ssl_certificate=` component of rgw_frontends it looks for a cert on disk, otherwise it'll look for a config option 'rgw_ssl_cert' or something. Which I assume the config class will probe the mon config database for.
Anyway I can go work on a RGW patch for that and have a play.

liewegas commented:
> > Two high-level comments:
> > 
> > 1. I don't think we should be putting anything in the daemon's ceph.conf unless it is absolutely necessary.  Instead, we should put options in the mon's config database, and (in most cases) apply them to the entire set/zone of rgw daemons.  This puts config mostly out of scope for the orchestrator layer.
> 
> I thought I was adding everything to the mon config database. I don't touch the ceph.conf added to the container. Though I'm probably just misunderstanding what your saying.

Oh I see, I didn't look at the code that closely.  I think the main point is that everything that's in that  json file (except teh crt)  should go into the mon config db (ceph config set ...).  AFAICS there's no reason for the orchestrator to have any understanding or plumbing for anything about rgw except which host it will go on and what it will be named.  I think there is a lot of cleanup we can/should do in the current orchestrator.py spec!

> > 1. IMO the SSL cert should be stored on the mon's config-key database.  We should teach radosgw how to fetch it from there when it starts up instead of copying it onto each node.
> 
> oh interesting to put it in via `ceph config ...` and then in rgw look for the cert data there. That's kinda cool. How about if RGW sees the `ssl_certificate=` component of rgw_frontends it looks for a cert on disk, otherwise it'll look for a config option 'rgw_ssl_cert' or something. Which I assume the config class will probe the mon config database for.
> Anyway I can go work on a RGW patch for that and have a play.

For the cert, it should go on the 'config-key' storage area (ceph config-key dump, ceph config-key get, etc.).  I think the next step is a patch to RGW that teaches it how to load the cert from there (instead of from a file).  That would be a great next step, I think!

matthewoliver commented:
Thanks @liewegas for the clarification, and introducing me to `ceph config-key`, will have a play with that today :)

Also done the rebase.

matthewoliver commented:
Lol, the batch commit of the github PR api I guess isn't adding the signed-off-by's, probably righly so. So will fix this when I squash all the patches together :)

mgfritch commented on src/cephadm/cephadm:
```diff
@@ -1000,8 +1000,8 @@ def get_config_and_keyring():
-def get_config_and_both_keyrings():
-    # type: () -> Tuple[str, str, Optional[str]]
+def get_config_keyrings_and_files():
+     # type: () -> Tuple[str, str, Optional[str], Dict]
```
Somehow got an extra space in there .. 
```suggestion
    # type: () -> Tuple[str, str, Optional[str], Dict]
```

mgfritch commented on src/cephadm/cephadm:
```diff
@@ -1012,7 +1012,16 @@ def get_config_and_both_keyrings():
+            config = d.pop('config')
+        if d.get('keyring'):
+            keyring = d.pop('keyring')
+        if d.get('crash_keyring'):
```
Remove this block in favor of the logic below.... the multiline code suggestions in github didn't work right :frowning_face:

matthewoliver replied:
no it didn't. Thanks github.

mgfritch replied:
indeed, thanks github :rofl:

matthewoliver commented:
missed the space.. fixed now

cbodley commented:
> 2. IMO the SSL cert should be stored on the mon's config-key database.  We should teach radosgw how to fetch it from there when it starts up instead of copying it onto each node.

i don't really have an intuition here, bringing @mdw-at-linuxbox in for feedback

mdw-at-linuxbox commented:
1/ you need to store the private key for that cert somewhere too.
2/ the key should not be exposed on the wire.
3/ in the overall workflow, the key needs to be generated, then a cert request, then a (potentially long) delay to get the actual cert (from a 3rd party), then ceph can be given the key + cert.
4/ in a large deployment, you might have multiple rgw's with different certs.  This might be for something complicated like a multi-zone setup, or something simple such as partitioning load to take better advantage of caching between different uses.  If you're storing certificates in some common location, you will want some method of indirection to allow this, and not one global certificate.
also,
5/ there is another overall workflow that has to do with renewing certificates.  Clever sites might have existing automation that can push certificates and keys.

matthewoliver commented:
> 1/ you need to store the private key for that cert somewhere too.
oh yeah, I can plumb that through too.

> 2/ the key should not be exposed on the wire.
good point. What is your definition of on the wire? is this a reason against using the mon config store? Or do we assume messenger v2 being encrypted is ok here?

Or is it sending the key through stdin through the ceph cli > orchestrator/daemon > cephadm.. though the last hop is via SSH isn't it, so that's encrypted.

If it's just the cli > daemon part of the leg, then don't we have the same problem with using the mon config store? or are we encrypting the cli traffic too? or at least the json we send. I guess I could just go check this.

> 3/ in the overall workflow, the key needs to be generated, then a cert request, then a (potentially long) delay to get the actual cert (from a 3rd party), then ceph can be given the key + cert.

I'd have throught this would happen externally from ceph. Ie. we just need to provide ceph the key + cert.
An advantage of using the store is there is a single place to dump the cert and key so will make the renewing process easier.

> 4/ in a large deployment, you might have multiple rgw's with different certs. This might be for something complicated like a multi-zone setup, or something simple such as partitioning load to take better advantage of caching between different uses. If you're storing certificates in some common location, you will want some method of indirection to allow this, and not one global certificate.
> also,

For multizone envs or even just giving RGW servers different certs, would it make more sense to use the config store, as apposed to the config-key store, when you can define the who. Thereby providing a 'who'. Ie globally client, a realm or zone.
or is there probably a limit to value stored in the config store.

I'm assuming here, that the config-key store doesn't have this granularity. We could do some playing with key names having realm or zone prefixed or some such, but that's starting to sound a little too hacky.

> 5/ there is another overall workflow that has to do with renewing certificates. Clever sites might have existing automation that can push certificates and keys.

So long as we support the use of files, or these clever sites can just `ceph config-key set <key> <cert_data>` and be done.

matthewoliver commented on src/pybind/mgr/cephadm/module.py:
```diff
@@ -1756,6 +1765,12 @@ def add_rgw(self, spec):
+            'who': 'client.rgw.' + spec.name,
+            'name': 'rgw_frontends',
+            'value': spec.rgw_frontend_config_value(),
+        })
```
While testing this, I don't see RGW picking up this rgw_frontends config option. Is there something I'm missing in setting this up? I assumed the ConfigProxy would search the mon config store, is this not the case?

I guess I could debug to find out.

matthewoliver replied:
Yeah, I think this is currently the major problem in the patch. As far as I can see, RGW doesn't look for the rgw_frontends value in the mon config store. I'd just assumed this was built into the the ConfigProxy.

Should we change @yehudasa 's https://github.com/ceph/ceph/pull/33287 to make the get keys from config-key available for other RGW values. And store the rgw_frontends there too (I guess check if it's missing in ceph.conf and  before defaulting to the default frontend configuration)?

ATM the only way to get this `rgw_frontends` working as it is, with out another RGW change is to write it into the ceph.conf in the container.. which I'd really rather not do.

Thoughts?

matthewoliver replied:
But if that is correct (not search config store) then the rgw_realm and rgw_zone is also working above.

I wonder how hard it would be to add mon config support to ConfigProxy.. though I guess continual calls to the network could be problematic. Or does this actaully work and I've just missed something obvious?

yehudasa replied:
@matthewoliver there is a problem with rgw order of initialization that rgw_frontends happens before it even connects to the cluster (it seems like it). I'm not sure that changing it won't break something else. In any case I'm working on adding a default frontends configurable that will allow using the config-key the way @liewegas suggested.

yehudasa replied:
@matthewoliver my PR now added default ssl cert/key configuration for the beast frontend via config-key.

matthewoliver replied:
Awesome! I'll rebase this on your PR on my dev machine and rework this one based on it. That way we can give them both a test.

matthewoliver commented:
Added the ability to define an SSL key file as well. Still not sure if this is the best way to pass the key file.

sebastian-philipp commented:
> is the mon integration going to be limited to just some bootstrap configurations?

Limiting this to just bootstrapping would make it kind of useless for cephadm's needs.

> I'm not sure it's going to be ideal for the stage where deploying a new application may bring new DNS naming and SSL cert material?

mattbenjamin commented:
@sebastian-philipp but what is the actual scope and intent of cephadm in this?  is it in fact sensible to have new secrets moving into the ceph mon or ceph config to match DNS common names being served by RGW?  today, we have some one-off override style params in the Ceph conf for setting up a global DNS name or keystone server or whatever, but the actual dynamic setup (e.g., zones, as Marcus mentioned) are all represented as objects in RGW's metadata

sebastian-philipp commented:
> @sebastian-philipp but what is the actual scope and intent of cephadm in this?

The goal is to make cephadm enterprise-ready. Day 1 and day 2. 

> Is it in fact sensible to have new secrets moving into the ceph mon or ceph config to match DNS common names being served by RGW?

Um, lack of knowledge on my side. I'll have to get back to you on this.

> ...

yehudasa commented:
@matthewoliver @liewegas @mattbenjamin @cbodley Whatever the the decision as to what would be the best or correct approach, here's a config via `config-key` of beast frontend:
https://github.com/ceph/ceph/pull/33287

matthewoliver commented:
Awesome work! Thanks so much for this, however we move forward, this will
be an awesome help!

KK, so we actually run commands to the mon. Good to know. I would've
expected the configproxy to grow the smarts to look in the mon store, so it
would be seamless to all the listeners. Anyway, I see we use mon_commands
elsewhere, so sweet. Again thanks heaps! You rock!

Matt

On Fri, Feb 14, 2020 at 7:46 AM Yehuda Sadeh <notifications@github.com>
wrote:

> @matthewoliver <https://github.com/matthewoliver> @liewegas
> <https://github.com/liewegas> @mattbenjamin
> <https://github.com/mattbenjamin> @cbodley <https://github.com/cbodley>
> Whatever the the decision as to what would be the best or correct approach,
> here's a config via config-key of beast frontend:
> #33287 <https://github.com/ceph/ceph/pull/33287>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ceph/ceph/pull/32951?email_source=notifications&email_token=AAAK3S4MQPVVKK4K3PS27YLRCWWS3A5CNFSM4KM67YK2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELWRREY#issuecomment-585963667>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAK3S73DH5OR2JXOKZM6RDRCWWS3ANCNFSM4KM67YKQ>
> .
>

matthewoliver commented:
Just a rebase now that @yehudasa's patch has landed. And started reworking. Am just testing and debugging the changes now :)

sebastian-philipp commented:
needs rebase

matthewoliver commented:
Thanks :) I have a new version to push, but ran outta time today.
I'll push what I have first thing in the morning.

On Tue., 10 Mar. 2020, 10:38 pm Sebastian Wagner <notifications@github.com
wrote:

> needs rebase
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ceph/ceph/pull/32951?email_source=notifications&email_token=AAAK3S23UXJMKNBORN7OWW3RGYRDBA5CNFSM4KM67YK2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEOLB6EY#issuecomment-597040915>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAK3S4SYYF33K3Z53MIQ5LRGYRDBANCNFSM4KM67YKQ>
> .
>

sebastian-philipp commented on src/pybind/mgr/cephadm/module.py:
```diff
@@ -2474,6 +2482,35 @@ def _config_rgw(self, spec):
         })
 
+        if spec.rgw_frontend_ssl_certificate:
+            if isinstance(spec.rgw_frontend_ssl_certificate, (list, tuple)):
```
just wondering, how can this be a tuple?

matthewoliver replied:
well a tuple being more of an immutable list, so really just depends on how the data was sent in. This just covers bases, so if it's any "list" like data structure we'll handle it.

matthewoliver replied:
I'll remove the tuple so we can resolve this discussion, poke me if I forget ;)

sebastian-philipp commented on src/pybind/mgr/orchestrator/_interface.py:
```diff
@@ -1624,6 +1624,7 @@ def __init__(self,
         assert service_type
         self.service_type = service_type
         self.service_id = service_id
+        self.name = self.service_name()
```
do we need this?

matthewoliver replied:
I did. since I rebased this, I get the error what RGWSpec doesn't have the attribute name. Whether I use the json file or not. Couldn't see anything I was missing in the class.

Put it here, so every spec will have it.. though I guess I should go check that other Specs fire the error without it.

sebastian-philipp commented on src/pybind/mgr/orchestrator/module.py:
```diff
@@ -571,8 +571,8 @@ def _mds_add(self, fs_name, placement=None):
-        'name=realm_name,type=CephString '
-        'name=zone_name,type=CephString '
+        'name=realm_name,type=CephString,req=false '
+        'name=zone_name,type=CephString,req=false '
```
this change seems a bit unrelated to this PR?

matthewoliver replied:
Yes and no, this means we can pass: ` ceph orch add rgw -i <json>` with out having to pass in the realm and zone on the command line.

Now that, or rather when, we can use the config-key store then maybe the `-i` becomes out of scope for the PR.

sebastian-philipp commented on src/cephadm/cephadm:
```diff
@@ -1270,10 +1272,26 @@ def extract_uid_gid(img='', file_path='/var/lib/ceph'):
     (uid, gid) = out.split(' ')
     return (int(uid), int(gid))
 
+def deploy_extra_files(extra, data_dir, uid, gid):
```
as we now use the mon store to keep the cert and key, do we still need to deploy extra files at all right now?

matthewoliver replied:
yeah, we might be able to loose this. Still trying to get the thing running. ATM when I run this in my vstart env, the RGW in the container fails to talk to monitor config. 

Or rather, when I jump into the docker container and pass correct beast details it fails:

```
# /bin/radosgw -d -setuser ceph --setgroup ceph --rgw-region=default --rgw-zone=default --keyring=/var/lib/ceph/radosgw/ceph-rgw.default.default.ironic-moliver.rslfos/keyring --rgw_frontends=beast port=8888 ssl_port
=4443 ssl_certificate=config://rgw/cert/default/default.crt ssl_key=config://rgw/cert/default/default.crt -d                 
2020-03-12T00:30:59.367+0000 7ff527191700 -1 monclient(hunting): handle_auth_bad_method server allowed_methods [2] but i only support [2,1]
2020-03-12T00:30:59.367+0000 7ff526990700 -1 monclient(hunting): handle_auth_bad_method server allowed_methods [2] but i only support [2,1]
2020-03-12T00:30:59.367+0000 7ff52618f700 -1 monclient(hunting): handle_auth_bad_method server allowed_methods [2] but i only support [2,1]
failed to fetch mon config (--no-mon-config to skip)    
```

sebastian-philipp replied:
did you pass the path to the vstart ceph.conf?

sebastian-philipp commented:
needs rebase

sebastian-philipp commented:
ping?

matthewoliver commented:
Rebased. Because things have moved under the hood quite a bit in the rebase, will now spend sometime debugging to make sure it's working *crosses fingers*.

sebastian-philipp commented:
> Rebased. Because things have moved under the hood quite a bit in the rebase, will now spend sometime debugging to make sure it's working _crosses fingers_.

is there any way I can help you?

liewegas commented:
I still think this is making things overly complicated.  

Right now, you can use SSL without any orch changes just by putting the cert in config-key where radosgw will automatically find it.  The only downside to this is that the behavior isn't consistent with rook.  I think we have a couple options:

- modify rook to pull the cert out of config-key if it is not included in the ObjectStore.  users for cephadm and rook will put the cert in config-key and be done with it.
- expand the RGWSpec to include the ssl cert, and have cephadm store it in config-key, and have rook store it in a k8s secret or whatever.

This PR adds a bunch of extra code to copy the cert into the container, but I don't think this actually buys us anything.  I have a slight preference is probably the first option above, but I'm fuzzy on what the "normal" rook CR looks like... is it normally the k8s user's job to put the cert in a secret before setting the CR, for instance?  And what happens in k8s when you have a service in front of rgw... is rgw itself running unencrypted, with the SSL terminated by the service/load balancer/whatever?

This might also intersect with the idea of an orch abstractions for haproxy that generalizes load balancing.  In that case, there are presumably two ways to set this all up: have SSL terminate with  haproxy, so that the cert goes there, or SSL terminate with radosgw itself.  Both should probably(?) be allowed, and in the former case the cert wouldn't go into the radosgw config-key location.

matthewoliver commented:
Thanks everyone. The extra parts predate rgw getting stuff from the key store. Will start reworking it to suite the new state of the world.

matthewoliver commented:
OK, cleaned up the extra files stuff. Currently the plan is to build up the frontends with the `confing//key/path/for/cert/and/keys` if ssl is passed. If you use the json file to pass the cert and key then cephadm will insert them into the config-key.

This way I hope we can do both. Just pass ssl, or optionally also pass the key and cert.

At the moment, as the code currently stands, you can pass in the key/cert and ssl=False and ssl wont be enabled. But I think it probably makes more sense for the passing the key/certs in should override ssl to true.

I haven't tested this code. Pushing it up here so I can pull it down on my dev server, so testing will start soon. I suspect a few more updates are coming ;)

matthewoliver commented on src/cephadm/samples/rgw_ssl.json:
```diff
@@ -0,0 +1,103 @@
+    "hosts": [{
+      "hostname": "ironic-moliver",
+      "name": "ironic-moliver",
+      "network": "ironic-moliver"
```
ignore all this for now, will clean it up (or remove this sample) before we merge.

matthewoliver commented on src/pybind/mgr/cephadm/module.py:
```diff
@@ -2741,31 +2741,56 @@ def _config_rgw(self, spec):
+            'who': f"{utils.name_to_config_section('rgw')}.{spec.name}",
             'name': 'rgw_frontends',
-            'value': v,
+            'value': spec.rgw_frontend_config_value(),
```
hmm, we're now doing this twice, will fix up.

matthewoliver commented on src/python-common/ceph/deployment/service_spec.py:
```diff
@@ -502,6 +502,8 @@ def __init__(self,
                  subcluster=None,  # type: Optional[str]
                  rgw_frontend_port=None,  # type: Optional[int]
+                 rgw_frontend_ssl_certificate=None, # type Optional[List[str]]
+                 rgw_frontend_ssl_key=None, # type: Optional[List[str]]
```
ops, forgot to put these in variables for RGWSpec. Will be fixed in the next version

mgfritch replied:
`make check` is currently failing on this ...

sebastian-philipp commented on src/pybind/mgr/cephadm/module.py:
```diff
@@ -2741,31 +2741,49 @@ def _config_rgw(self, spec):
+            else:
+                cert_data = spec.rgw_frontend_ssl_certificate
+            ret, out, err = self.mon_command({
+                'prefix': 'config-key set',
+                'key': f'rgw/cert/{spec.rgw_realm}/{spec.rgw_zone}.crt',
+                'value': cert_data,
```
Do you know, if we need `who` as well here?

matthewoliver replied:
As far as I'm aware, `config` needs a `who`. But a `config-key` doesn't, is just key value pairs.

mgfritch commented on src/cephadm/samples/rgw_ssl.json:
```diff
@@ -0,0 +1,103 @@
+      "805lq+iy7Mkhb+dlr4R9WhMWDyGwgYs=",
+      "-----END PRIVATE KEY-----"
+  ]
+}
```
The samples in this dir are typically used with `bin/cephadm --config-json <sample>.json`, where this one appears to relate to `mgr/cephadm`?

should we move it to `mgr/orchestrator/samples` or `mgr/cehpadm/samples` ?

sebastian-philipp replied:
doc/cephadm and mention it in the official docs?

mgfritch commented:
@matthewoliver needs rebase

matthewoliver commented:
Rebased.

When I seem to deploy RGW with SSL I am now down to 3 "issues":
1. now that we default to port 80 and 443. RGW fails to bind to this address, as it doesn't have permission to bind to priv ports. Do we drop user perms before we setup (bind) the frontend in RGW or is this more a container level thing? (I guess I could try and run rgw as root in the container and see what happens).

2. When I specify the SSL cert and key in a json, the `config-key set` mon-command sends the file contents as a string value in the json. Ofter deploy the key value is ''. So need to figure out why? is there a limit to what we send. If I manaully use `ceph config-key set <key> -i <cert>` it's fine.

3. When I do deploy, and the certs are there, because I've added them manually and just use: `ceph orch daemon add rgw default default <host>` (on a non priv port). The daemon is then removed as an orphan daemon. So need to figure out why the orch doesn't like it. Maybe the service name or ID needs a tweak?

Anyway, getting very close. We could just remove the SSL cert/key from spec and json, and expect a user to push a cert/key into the config-key store.

sebastian-philipp commented:
> Rebased.
> 
> When I seem to deploy RGW with SSL I am now down to 3 "issues":
> 
> 1. now that we default to port 80 and 443. RGW fails to bind to this address, as it doesn't have permission to bind to priv ports. Do we drop user perms before we setup (bind) the frontend in RGW or is this more a container level thing? (I guess I could try and run rgw as root in the container and see what happens).

I don't know in detail how podman as root with `--net=host` and rgw as non-root within the container interacts with each other. 

> 2. When I specify the SSL cert and key in a json, the `config-key set` mon-command sends the file contents as a string value in the json. Ofter deploy the key value is ''. So need to figure out why? is there a limit to what we send. If I manaully use `ceph config-key set <key> -i <cert>` it's fine.

the mon command and `ceph config-key set ` is basically the same thing. Maybe you could `assert` that the cert is non-empty?

> 3. When I do deploy, and the certs are there, because I've added them manually and just use: `ceph orch daemon add rgw default default <host>` (on a non priv port). The daemon is then removed as an orphan daemon. So need to figure out why the orch doesn't like it. Maybe the service name or ID needs a tweak?

Related to https://tracker.ceph.com/issues/44934 ?
> 
> Anyway, getting very close.

> We could just remove the SSL cert/key from spec and json, and expect a user to push a cert/key into the config-key store.

Which might actually be a valid way for now. We just need some documentation for it somewhere. (and we need to document this anyway for non-cephadm clusters)

matthewoliver commented:
> > Rebased.
> > When I seem to deploy RGW with SSL I am now down to 3 "issues":
> > 
> > 1. now that we default to port 80 and 443. RGW fails to bind to this address, as it doesn't have permission to bind to priv ports. Do we drop user perms before we setup (bind) the frontend in RGW or is this more a container level thing? (I guess I could try and run rgw as root in the container and see what happens).
> 
> I don't know in detail how podman as root with `--net=host` and rgw as non-root within the container interacts with each other.

No idea, step 1 I'll run rgw as root just to confirm it isn't a different issue. Might add the ability to supply the port through the commandline, as it'll make it easier to test, at least temp. 

> 
> > 1. When I specify the SSL cert and key in a json, the `config-key set` mon-command sends the file contents as a string value in the json. Ofter deploy the key value is ''. So need to figure out why? is there a limit to what we send. If I manaully use `ceph config-key set <key> -i <cert>` it's fine.
> 
> the mon command and `ceph config-key set ` is basically the same thing. Maybe you could `assert` that the cert is non-empty?

yeah, when I print it out, it's defintely non empty. It's the cert as a string with a bunch of '\n'. 

> 
> > 1. When I do deploy, and the certs are there, because I've added them manually and just use: `ceph orch daemon add rgw default default <host>` (on a non priv port). The daemon is then removed as an orphan daemon. So need to figure out why the orch doesn't like it. Maybe the service name or ID needs a tweak?
> 
> Related to https://tracker.ceph.com/issues/44934 ?

YES!! This looks very similar. So I wont focus too hard on solving this as it seems you might have already. Though at least I can test your patch :)
> 
> > Anyway, getting very close.
> 
> > We could just remove the SSL cert/key from spec and json, and expect a user to push a cert/key into the config-key store.
> 
> Which might actually be a valid way for now. We just need some documentation for it somewhere. (and we need to document this anyway for non-cephadm clusters)

Thanks will continue pushing this. Will hopefully get something new up soon.

mgfritch commented on src/python-common/ceph/deployment/service_spec.py:
```diff
@@ -511,6 +511,8 @@ def __init__(self,
                  rgw_zone=None,  # type: Optional[str]
                  subcluster=None,  # type: Optional[str]
                  rgw_frontend_port=None,  # type: Optional[int]
+                 rgw_frontend_ssl_certificate=None, # type Optional[List[str]]
+                 rgw_frontend_ssl_key=None, # type: Optional[List[str]]
```
```suggestion
                 rgw_frontend_ssl_certificate=None,  # type Optional[List[str]]
                 rgw_frontend_ssl_key=None,  # type: Optional[List[str]]
```

```
ceph/deployment/service_spec.py:514:52: E261 at least two spaces before inline comment
ceph/deployment/service_spec.py:515:44: E261 at least two spaces before inline comment
```

matthewoliver commented:
OK this new version has:

 - working cert/ssl via json working
 - fixed another can't find cert/key error because we needed more maps on mon to execute commands
 - Added the spec to the spec cache so it wont get cleaned up as an orphan daemon.

So all working (well minus the bind on priv-port) which may or maynot be just my env.
